TestNewRouteCallsSyncHandler is flaky

<!--
/area API
/area networking
/kind bug
/kind good-first-issue
-->

We intermittently hit the following flake during our test runs:
https://gubernator.knative.dev/build/knative-prow/logs/ci-knative-serving-continuous/1061363294306045952

The error is:
```
logger.go:130: 2018-11-10T21:08:56.164Z	DEBUG	TestNewRouteCallsSyncHandler.route-controller	reconciler/reconciler.go:102	Creating event broadcaster	{"knative.dev/controller": "route-controller"}
```

More useful output:
```
I1110 21:09:16.529] === RUN   TestNewRouteCallsSyncHandler
I1110 21:09:16.529] --- FAIL: TestNewRouteCallsSyncHandler (3.02s)
I1110 21:09:16.529] 	logger.go:130: 2018-11-10T21:08:56.164Z	DEBUG	TestNewRouteCallsSyncHandler.route-controller	reconciler/reconciler.go:102	Creating event broadcaster	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.529] 	logger.go:130: 2018-11-10T21:08:56.172Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:119	Setting up event handlers	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.530] 	logger.go:130: 2018-11-10T21:08:56.173Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:158	Setting up ConfigMap receivers	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.530] 	logger.go:130: 2018-11-10T21:08:56.174Z	INFO	TestNewRouteCallsSyncHandler.route-controller.config-store	configmap/store.go:166	route config "config-domain" config was added or updated: &{map[test-domain.dev:0xc42049c0b0 prod-domain.com:0xc42049c0c0]}	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.530] 	logger.go:130: 2018-11-10T21:08:56.175Z	INFO	TestNewRouteCallsSyncHandler.route-controller.config-store	configmap/store.go:166	route config "config-gc" config was added or updated: &{24h0m0s 15h0m0s 1 5h0m0s}	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.530] 	logger.go:130: 2018-11-10T21:08:56.183Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:189	Starting controller and workers	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.530] 	logger.go:130: 2018-11-10T21:08:56.183Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:197	Started workers	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.531] 	logger.go:130: 2018-11-10T21:08:56.194Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:218	Reconciling route :&{{ } {test-route  test /apis/serving/v1alpha1/namespaces/test/Routes/test-route   0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[route:test-route] map[] [] nil [] } {0 [{ test-rev  100}]} {  <nil> [] [{AllTrafficAssigned Unknown Error {2018-11-10 21:08:56.194309248 +0000 UTC m=+0.098993373}  } {IngressReady Unknown Error {2018-11-10 21:08:56.194318213 +0000 UTC m=+0.099002322}  } {Ready Unknown Error {2018-11-10 21:08:56.194328851 +0000 UTC m=+0.099012960}  }] 0}}	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1110 21:09:16.531] 	logger.go:130: 2018-11-10T21:08:56.195Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:266	Successfully synced "test/test-route"	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.531] 	logger.go:130: 2018-11-10T21:08:56.201Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:218	Reconciling route :&{{ } {test-route  test /apis/serving/v1alpha1/namespaces/test/Routes/test-route   0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[route:test-route] map[] [] nil [] } {0 [{ test-rev  100}]} {  <nil> [] [{AllTrafficAssigned False Error {2018-11-10 21:08:56.194987736 +0000 UTC m=+0.099671880} RevisionMissing Revision "test-rev" referenced in traffic not found.} {IngressReady Unknown Error {2018-11-10 21:08:56.194318213 +0000 UTC m=+0.099002322}  } {Ready False Error {2018-11-10 21:08:56.195023824 +0000 UTC m=+0.099707948} RevisionMissing Revision "test-rev" referenced in traffic not found.}] 0}}	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1110 21:09:16.531] 	logger.go:130: 2018-11-10T21:08:56.202Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:266	Successfully synced "test/test-route"	{"knative.dev/controller": "route-controller"}
I1110 21:09:16.531] 	queueing_test.go:137: Timed out waiting for hooks to complete
I1110 21:09:16.532] 	logger.go:130: 2018-11-10T21:08:59.178Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:199	Shutting down workers	{"knative.dev/controller": "route-controller"}
```

You can see that it clearly times out, however, the Route is successfully synced twice in the first ~50ms after the test begins.

So far, I have been unable to reproduce this locally running it `10000` times normally, or `1000` with `-race`.

 cc @adrcunha for any ideas on how what I might be doing is different from Prow.

I just hit this in https://github.com/knative/serving/pull/2724

[logs](https://gubernator.knative.dev/build/knative-prow/pr-logs/pull/knative_serving/2724/pull-knative-serving-unit-tests/1073680957363458049/):
```
I1214 20:50:15.503] === RUN   TestNewRouteCallsSyncHandler
I1214 20:50:15.503] --- FAIL: TestNewRouteCallsSyncHandler (3.04s)
I1214 20:50:15.503] 	logger.go:130: 2018-12-14T20:49:52.954Z	DEBUG	TestNewRouteCallsSyncHandler.route-controller	reconciler/reconciler.go:105	Creating event broadcaster	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.503] 	logger.go:130: 2018-12-14T20:49:52.961Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:119	Setting up event handlers	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.504] 	logger.go:130: 2018-12-14T20:49:52.963Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:158	Setting up ConfigMap receivers	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.504] 	logger.go:130: 2018-12-14T20:49:52.964Z	INFO	TestNewRouteCallsSyncHandler.route-controller.config-store	configmap/store.go:166	route config "config-gc" config was added or updated: &{24h0m0s 15h0m0s 1 5h0m0s}	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.504] 	logger.go:130: 2018-12-14T20:49:52.966Z	INFO	TestNewRouteCallsSyncHandler.route-controller.config-store	configmap/store.go:166	route config "config-domain" config was added or updated: &{map[test-domain.dev:0xc4200f60d8 prod-domain.com:0xc4200f60e8]}	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.504] 	logger.go:130: 2018-12-14T20:49:52.991Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:215	Starting controller and workers	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.505] 	logger.go:130: 2018-12-14T20:49:52.992Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:223	Started workers	{"knative.dev/controller": "route-controller"}
I1214 20:50:15.505] 	logger.go:130: 2018-12-14T20:49:53.078Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:225	Reconciling route :&{{ } {test-route  test /apis/serving/v1alpha1/namespaces/test/Routes/test-route   0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[route:test-route] map[] [] nil [] } {0 [{ test-rev  100}]} {  <nil> [] [{AllTrafficAssigned Unknown Error {2018-12-14 20:49:53.078025585 +0000 UTC m=+0.470937833}  } {IngressReady Unknown Error {2018-12-14 20:49:53.078043678 +0000 UTC m=+0.470955899}  } {Ready Unknown Error {2018-12-14 20:49:53.078153216 +0000 UTC m=+0.471065455}  }] 0}}	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1214 20:50:15.505] 	logger.go:130: 2018-12-14T20:49:53.079Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:275	Reconcile succeeded. Time taken: 1.844355ms.	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1214 20:50:15.506] 	logger.go:130: 2018-12-14T20:49:53.086Z	INFO	TestNewRouteCallsSyncHandler.route-controller	route/route.go:225	Reconciling route :&{{ } {test-route  test /apis/serving/v1alpha1/namespaces/test/Routes/test-route   0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[route:test-route] map[] [] nil [] } {0 [{ test-rev  100}]} {  <nil> [] [{AllTrafficAssigned False Error {2018-12-14 20:49:53.079134367 +0000 UTC m=+0.472046647} RevisionMissing Revision "test-rev" referenced in traffic not found.} {IngressReady Unknown Error {2018-12-14 20:49:53.078043678 +0000 UTC m=+0.470955899}  } {Ready False Error {2018-12-14 20:49:53.079182226 +0000 UTC m=+0.472094475} RevisionMissing Revision "test-rev" referenced in traffic not found.}] 0}}	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1214 20:50:15.506] 	logger.go:130: 2018-12-14T20:49:53.087Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:275	Reconcile succeeded. Time taken: 1.705554ms.	{"knative.dev/controller": "route-controller", "knative.dev/key": "test/test-route"}
I1214 20:50:15.506] 	queueing_test.go:140: Timed out waiting for hooks to complete
I1214 20:50:15.506] 	logger.go:130: 2018-12-14T20:49:55.985Z	INFO	TestNewRouteCallsSyncHandler.route-controller	controller/controller.go:225	Shutting down workers	{"knative.dev/controller": "route-controller"}
```

--- a/pkg/reconciler/v1alpha1/revision/revision_test.go
+++ b/pkg/reconciler/v1alpha1/revision/revision_test.go
@@ -725,6 +725,7 @@ func TestGlobalResyncOnConfigMapUpdate(t *testing.T) {
 	}}
 
 	for _, test := range tests {
+		test := test
 		t.Run(test.name, func(t *testing.T) {
 			controllerConfig := getTestControllerConfig()
 			kubeClient, servingClient, _, _, controller, kubeInformer, servingInformer, cachingInformer, watcher, _ := newTestControllerWithConfig(t, controllerConfig)
@@ -759,6 +760,10 @@ func TestGlobalResyncOnConfigMapUpdate(t *testing.T) {
 			servingInformer.Start(stopCh)
 			kubeInformer.Start(stopCh)
 			cachingInformer.Start(stopCh)
+
+			servingInformer.WaitForCacheSync(stopCh)
+			kubeInformer.WaitForCacheSync(stopCh)
+			cachingInformer.WaitForCacheSync(stopCh)
 
 			if err := watcher.Start(stopCh); err != nil {
 				t.Fatalf("Failed to start configuration manager: %v", err)

--- a/pkg/reconciler/v1alpha1/route/queueing_test.go
+++ b/pkg/reconciler/v1alpha1/route/queueing_test.go
@@ -123,6 +123,9 @@ func TestNewRouteCallsSyncHandler(t *testing.T) {
 	servingInformer.Start(stopCh)
 	configMapWatcher.Start(stopCh)
 
+	kubeInformer.WaitForCacheSync(stopCh)
+	servingInformer.WaitForCacheSync(stopCh)
+
 	// Run the controller.
 	eg.Go(func() error {
 		return controller.Run(2, stopCh)

--- a/pkg/reconciler/v1alpha1/route/route_test.go
+++ b/pkg/reconciler/v1alpha1/route/route_test.go
@@ -928,6 +928,7 @@ func TestGlobalResyncOnUpdateDomainConfigMap(t *testing.T) {
 	}}
 
 	for _, test := range tests {
+		test := test
 		t.Run(test.expectedDomainSuffix, func(t *testing.T) {
 			_, servingClient, controller, _, kubeInformer, servingInformer, watcher := newTestSetup(t)
 
@@ -952,6 +953,10 @@ func TestGlobalResyncOnUpdateDomainConfigMap(t *testing.T) {
 
 			servingInformer.Start(stopCh)
 			kubeInformer.Start(stopCh)
+
+			servingInformer.WaitForCacheSync(stopCh)
+			kubeInformer.WaitForCacheSync(stopCh)
+
 			if err := watcher.Start(stopCh); err != nil {
 				t.Fatalf("failed to start configuration manager: %v", err)
 			}
@@ -962,8 +967,7 @@ func TestGlobalResyncOnUpdateDomainConfigMap(t *testing.T) {
 			route := getTestRouteWithTrafficTargets([]v1alpha1.TrafficTarget{})
 			route.Labels = map[string]string{"app": "prod"}
 
-			routeClient := servingClient.ServingV1alpha1().Routes(route.Namespace)
-			routeClient.Create(route)
+			servingClient.ServingV1alpha1().Routes(route.Namespace).Create(route)
 
 			test.doThings(watcher)
 

